osquery/config/parsers/decorators.cpp

```diff
@@ -0,0 +1,246 @@
+                         const pt::ptree& decorators);
+
+  /// Clear the decorations created from decorators for the given source.
+  void clearSources(const std::string& source);
+
+ public:
+  /// Set of configuration sources to the set of decorator queries.
+  std::map<std::string, std::vector<std::string>> always_;
+
+  /// Set of configuration sources to the set of on-load decorator queries.
+  std::map<std::string, std::vector<std::string>> load_;
+
+  /// Set of configuration sources to valid intervals.
+  std::map<std::string, std::map<size_t, std::vector<std::string>>> intervals_;
+
+ public:
+  /// The result set of decorations, column names and their values.
+  static DecorationStore kDecorations;
+
+  /// Protect additions to the decorator set.
+  static Mutex kDecorationsMutex;
+};
+
+DecorationStore DecoratorsConfigParserPlugin::kDecorations;
+Mutex DecoratorsConfigParserPlugin::kDecorationsMutex;
+
+Status DecoratorsConfigParserPlugin::setUp() {
+  // Decorators are kept within customized data structures.
+  // No need to define a key for the ::getData API.
+  return Status(0, "OK");
+}
+
+Status DecoratorsConfigParserPlugin::update(const std::string& source,
+                                            const ParserConfig& config) {
+  clearSources(source);
+  clearDecorations(source);
+  if (config.count(PARSER_NAME) > 0) {
+    // Each of these methods acquires the decorator lock separately.
+    // The run decorators method is designed to have call sites throughout
+    // the code base.
+    updateDecorations(source, config.at(PARSER_NAME));
+    runDecorators(DECORATE_LOAD, 0, source);
+  }
+
+  return Status(0, "OK");
+}
+
+void DecoratorsConfigParserPlugin::clearSources(const std::string& source) {
+  // Reset the internal data store.
+  WriteLock lock(DecoratorsConfigParserPlugin::kDecorationsMutex);
+  intervals_[source].clear();
+  always_[source].clear();
+  load_[source].clear();
+}
+
+void DecoratorsConfigParserPlugin::updateDecorations(
+    const std::string& source, const pt::ptree& decorators) {
+  WriteLock lock(DecoratorsConfigParserPlugin::kDecorationsMutex);
+  // Assign load decorators.
+  auto& load_key = kDecorationPointKeys.at(DECORATE_LOAD);
+  if (decorators.count(load_key) > 0) {
+    for (const auto& item : decorators.get_child(load_key)) {
+      load_[source].push_back(item.second.data());
+    }
+  }
+
+  // Assign always decorators.
+  auto& always_key = kDecorationPointKeys.at(DECORATE_ALWAYS);
+  if (decorators.count(always_key) > 0) {
+    for (const auto& item : decorators.get_child(always_key)) {
+      always_[source].push_back(item.second.data());
+    }
+  }
+
+  // Check if intervals are defined.
+  auto& interval_key = kDecorationPointKeys.at(DECORATE_INTERVAL);
+  if (decorators.count(interval_key) > 0) {
+    auto& interval = decorators.get_child(interval_key);
+    for (const auto& item : interval) {
+      size_t rate = std::stoll(item.first);
+      if (rate % 60 != 0) {
```

Comment:
If you schedule your decorator to run once an hour (3600 seconds), it will pass this without warning. I can understand `if (rate < 60)`, but if I wanted to schedule something like every 4000 seconds, this would still fail. Intentional?